pyem/metadata.py

```diff
@@ -292,23 +292,82 @@ def cryosparc_2_cs_ctf_parameters(cs, df=None):
     log = logging.getLogger('root')
     if df is None:
         df = pd.DataFrame()
-    if 'ctf/tilt_A' in cs.dtype.names:
-        log.debug("Recovering beam tilt and converting to mrad")
-        df[star.Relion.BEAMTILTX] = np.arcsin(cs['ctf/tilt_A'][:, 0] / cs['ctf/cs_mm'] * 1e-7) * 1e3
-        df[star.Relion.BEAMTILTY] = np.arcsin(cs['ctf/tilt_A'][:, 1] / cs['ctf/cs_mm'] * 1e-7) * 1e3
-    if 'ctf/shift_A' in cs.dtype.names:
-        pass
-    if 'ctf/trefoil_A' in cs.dtype.names:
-        pass
-        # df[star.Relion.ODDZERNIKE] = cs['ctf/trefoil_A']
-    if 'ctf/tetrafoil_A' in cs.dtype.names:
-        pass
-        # df[star.Relion.EVENZERNIKE] = cs['ctf/tetra_A']
+    if 'ctf/accel_kv' in cs.dtype.names:
+        # Calculate electron wavelentgh
+        kv = cs['ctf/accel_kv']
+        wavelength = 12.2643247 / np.sqrt(kv * (1 + (kv * 0.978466e-6)))
+    else:
+        wavelength = None
     if 'ctf/anisomag' in cs.dtype.names:
         df[star.Relion.MAGMAT00] = cs['ctf/anisomag'][:, 0]
         df[star.Relion.MAGMAT01] = cs['ctf/anisomag'][:, 1]
         df[star.Relion.MAGMAT10] = cs['ctf/anisomag'][:, 2]
         df[star.Relion.MAGMAT11] = cs['ctf/anisomag'][:, 3]
+    if 'ctf/tilt_A' in cs.dtype.names:
+        log.debug("Recovering beam tilt and converting to mrad")
+        df[star.Relion.BEAMTILTX] = np.arcsin(cs['ctf/tilt_A'][:, 0] / cs['ctf/cs_mm'] * 1e-7) * 1e3
+        df[star.Relion.BEAMTILTY] = np.arcsin(cs['ctf/tilt_A'][:, 1] / cs['ctf/cs_mm'] * 1e-7) * 1e3
+        df[star.Relion.Z_neg1_3] = df[star.Relion.BEAMTILTX]  # does this also require a conversion factor of 2π * Cs * λ^2
+        df[star.Relion.Z_pos1_3] = df[star.Relion.BEAMTILTY]  # does this also require a conversion factor of 2π * Cs * λ^2
+    if 'ctf/shift_A' in cs.dtype.names:
+        df[star.Relion.Z_neg1_1] = (2 * np.pi) + cs['ctf/shift_A'][:, 0]
+        df[star.Relion.Z_pos1_1]  = (2 * np.pi) + cs['ctf/shift_A'][:, 1]
+    if 'ctf/trefoil_A' in cs.dtype.names and 'ctf/cs_mm' in cs.dtype.names and wavelength is not None:
+        unit_conversion_even = (2*np.pi) * cs['ctf/cs_mm'] * (wavelength**2)
+        df[star.Relion.Z_neg3_3] = unit_conversion_even * cs['ctf/trefoil_A'][:, 0]
+        df[star.Relion.Z_pos3_3] = unit_conversion_even * cs['ctf/trefoil_A'][:, 1]
+    if 'ctf/amp_contrast' in cs.dtype.names:
+        df[star.Relion.Z_0_0] = cs['ctf/amp_contrast'] - np.arccos(phi)  # what is phi? Could it be cs['ctf/phas_shift_rad'] ?
```

Comment:
Moving the discussion about phi to the line of code it belongs to...
> You're also right about phi, that is the phase shift nominally from using a phase plate during imaging. The amplitude contrast and the phase-plate shift both work the same though; estimating phase shifts is essentially estimating amplitude contrast per-micrograph.
>
> The amplitude contrast is a pure number from 0 to 1, since the phase shift is given in radians arccos is used to combine.

Reply:
So `phi = cs['ctf/phase_shift_rad']`? That seems most likely to be right, so I'll update the code with it.